### ISSUE TITLE
[temp.arg.template] p4 Saying PP is an id-expression is a wrong assumption

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5372]
The second bullet in [temp.arg.template] p4 says
> If PP declares a template parameter pack, then AA is the pack expansion PP... ([[temp.variadic]](https://eel.is/c++draft/temp.variadic)); otherwise, AA is the **[id-expression](https://eel.is/c++draft/expr.prim.id.general#nt:id-expression) PP**[.](https://eel.is/c++draft/temp.arg.template#4.2.sentence-2)  

The id-expression implies that PP can only be a non-type template parameter since whose declarator-id is an id-expression. The name of the type template parameter is either *typedef-name* or *template-name*, as per [temp.param] p3  

> A type-parameter whose identifier does not follow an ellipsis defines its identifier to be a typedef-name (if declared without template) or template-name (if declared with template) in the scope of the template declaration.  

Maybe, we change the last sentence to
> otherwise, AA is the identifier naming PP.

is better?

**Comment by xmh0511** [comment]
> For example, for declarator-id in [dcl.decl.general].

Yes, since the *declarator-id* is an *id-expression* on **grammar**. That is why I said *id-expression* may limit the template parameter as a non-type template parameter that has a declarator-id. However, any **type** template parameter is introduced by an **identifier**, the lookup rules apply to the name shouldn't interpret that name as an **id-expression**.

**Comment by xmh0511** [comment]
For template argument corresponding to template template-parameter, we have [temp.arg.template] p1 that specifies that the template-argument is expressed as *id-expression*. So, it could be arguably correct that we say `AA` is an id-expression when `PP` is a **template template-parameter**

However, we do not have a similar special phrase that specifies the type template-argument is expressed as an id-expression. [temp.arg.type] p1 says such a template-argument **shall** be a [type-id](https://eel.is/c++draft/dcl.name#nt:type-id).    

> A template-argument for a template-parameter which is a type shall be a type-id.

*type-id* is not an *id-expression* and vice versa.  

***
BTW, [temp.arg.template] p1 requires that the template-argument for **template template-paramter**  **shall** be the name of a **class template** or an **alias template**.  It is unclear whether a name introduced by a template template-parameter can be considered to be a name of a class template or of an alias template or not? We do not have a relevant phrase.

**Comment by jensmaurer** [comment]
For the last point, I think we want [temp.param] p3 to say "... _template-name_ that designates a template whose specializations are types".

**Comment by xmh0511** [comment]
> I think we want [temp.param] p3 to say "... template-name that designates a template whose specializations are types".

Is it [temp.arg.template] p1 instead of [temp.param] p3? [temp.arg.template] p1 might be changed to  

> A template-argument for a template template-parameter shall be a template-name that designates a template whose specializations are types, expressed as id-expression.   

***

How about the point that a *type-id*(especially, *typedef-name*) is not an *id-expression*. I think it also should be fixed.

**Comment by languagelawyer** [comment]
FWIW, [temp.arg.template]/1 says:
> A [template-argument](https://eel.is/c++draft/temp.names#nt:template-argument) for a template [template-parameter](https://eel.is/c++draft/temp.param#nt:template-parameter) shall be the name of a class template or an alias template, **expressed as** [id-expression](https://eel.is/c++draft/expr.prim.id.general#nt:id-expression)[.](https://eel.is/c++draft/temp.arg.template#1.sentence-1)

**Comment by jensmaurer** [comment]
[temp.arg.template] is concerned with template arguments, i.e. the use of templates. [temp.param] is concerned with template parameters, i.e. the declaration and definition of templates. In order to properly syntax-check a template definition, I think we need to know that the _identifier_ introduced via a template template parameter is actually (similar to) a class/alias template. Maybe we want to even say that those are alias templates.

I think [temp.arg.template] is fine as-is; this is just another place where _id-expression_ is used with a non-expression meaning (which ought to be fixed, but not like this).

**Comment by xmh0511** [comment]
@jensmaurer 
> a template template parameter is actually (similar to) a class/alias template.

I also think so. There are two remaining issues,  [temp.param] p3 might be fixed by: 
> - A type-parameter whose identifier does not follow an ellipsis declares its identifier to be an **alias template** that is a *template-name*(if declared with template)


Another issue is we should give a special rule to specify the name of a **type** template parameter (**declared without template**) whose name becomes a **typedef-name** is interpreted as an *id-expression* whenever it is used as a template-argument.
